lib/web/sri/sri.js

```diff
@@ -0,0 +1,302 @@
+ * "sha384", "sha512" » (corresponding to SHA-256, SHA-384, and SHA-512
+ * respectively). The ordering of this set is meaningful, with stronger
+ * algorithms appearing later in the set.
+ *
+ * @see https://w3c.github.io/webappsec-subresource-integrity/#valid-sri-hash-algorithm-token-set
+ */
+const validSRIHashAlgorithmTokenSet = ['sha256', 'sha384', 'sha512']
+
+// https://nodejs.org/api/crypto.html#determining-if-crypto-support-is-unavailable
+/** @type {import('crypto')} */
+let crypto
+try {
+  crypto = require('node:crypto')
+  const cryptoHashes = crypto.getHashes()
+
+  // If no hashes are available, we cannot support SRI.
+  if (cryptoHashes.length === 0) {
+    validSRIHashAlgorithmTokenSet.length = 0
+  }
+
+  for (let i = 0; i < validSRIHashAlgorithmTokenSet.length; i++) {
+    const algorithm = validSRIHashAlgorithmTokenSet[i]
+    // If the algorithm is not supported, remove it from the list.
+    if (cryptoHashes.includes(algorithm) === false) {
+      validSRIHashAlgorithmTokenSet.splice(i, 1)
+      i--
+    }
+  }
+  /* c8 ignore next 4 */
+} catch {
+  // If crypto is not available, we cannot support SRI.
+  validSRIHashAlgorithmTokenSet.length = 0
+}
+
+/**
+ * @param {*} algorithm
+ * @returns {algorithm is SRIHashAlgorithm}
+ */
+function isValidSRIHashAlgorithm (algorithm) {
+  return validSRIHashAlgorithmTokenSet.includes(algorithm)
+}
+
+/**
+ * @param {Uint8Array} bytes
+ * @param {string} metadataList
+ * @returns {boolean}
+ *
+ * @see https://w3c.github.io/webappsec-subresource-integrity/#does-response-match-metadatalist
+ */
+function bytesMatch (bytes, metadataList) {
+  // If node is not built with OpenSSL support, we cannot check
+  // a request's integrity, so allow it by default (the spec will
+  // allow requests if an invalid hash is given, as precedence).
+  /* istanbul ignore if: only if node is built with --without-ssl */
+  if (crypto === undefined) {
+    return true
+  }
+
+  // 1. Let parsedMetadata be the result of parsing metadataList.
+  const parsedMetadata = parseMetadata(metadataList)
+
+  // 2. If parsedMetadata is empty set, return true.
+  if (parsedMetadata.length === 0) {
+    return true
+  }
+
+  // 3. Let metadata be the result of getting the strongest
+  //    metadata from parsedMetadata.
+  const metadata = getStrongestMetadata(parsedMetadata)
+
+  // 4. For each item in metadata:
+  for (const item of metadata) {
+    // 1. Let algorithm be the item["alg"].
+    const algorithm = item.alg
+
+    // 2. Let expectedValue be the item["val"].
+    const expectedValue = item.val
+
+    // See https://github.com/web-platform-tests/wpt/commit/e4c5cc7a5e48093220528dfdd1c4012dc3837a0e
+    // "be liberal with padding". This is annoying, and it's not even in the spec.
+
+    // 3. Let actualValue be the result of applying algorithm to bytes .
+    const actualValue = applyAlgorithmToBytes(algorithm, bytes)
+
+    // 4. If actualValue is a case-sensitive match for expectedValue,
```

Comment:
The spec reads as though "case-sensitive match" is just an ===. It feels like we're missing something if we have to implement our own "base64 === base64url" function.